core/src/main/java/hivemall/utils/math/MatrixUtils.java

```diff
@@ -493,4 +498,202 @@ public static RealMatrix solve(@Nonnull final RealMatrix L, @Nonnull final RealM
+                "T and Qt must be the same shape");
+
+        // initial R = T
+        R.setSubMatrix(T.getData(), 0, 0);
+
+        // initial Qt = identity
+        Qt.setSubMatrix(new double[n][n], 0, 0);
+        for (int i = 0; i < n; i++) {
+            Qt.setEntry(i, i, 1);
+        }
+
+        for (int i = 0; i < n - 1; i++) {
+            // Householder projection for a vector x
+            // https://en.wikipedia.org/wiki/Householder_transformation
+            RealVector x = T.getSubMatrix(i, i + 1, i, i).getColumnVector(0);
+
+            double x0 = x.getEntry(0);
+            double sign = 0.d;
+            if (x0 < 0.d) {
+                sign = -1.d;
+            } else if (x0 > 0.d) {
+                sign = 1.d;
+            }
+
+            x.setEntry(0, x0 + sign * x.getNorm());
+            x = x.unitVector();
+
+            RealMatrix subR = R.getSubMatrix(i, i + 1, 0, n - 1);
+            R.setSubMatrix(subR.subtract(x.outerProduct(subR.preMultiply(x)).scalarMultiply(2)).getData(), i, 0);
+
+            RealMatrix subQt = Qt.getSubMatrix(i, i + 1, 0, n - 1);
+            Qt.setSubMatrix(subQt.subtract(x.outerProduct(subQt.preMultiply(x)).scalarMultiply(2)).getData(), i, 0);
+        }
+    }
+
+    /**
+     * Find eigenvalues and eigenvectors of given tridiagonal matrix T.
+     *
+     * http://web.csulb.edu/~tgao/math423/s94.pdf
+     * http://stats.stackexchange.com/questions/20643/finding-matrix-eigenvectors-using-qr-decomposition
+     *
+     * @param T target tridiagonal matrix
+     * @param nIter number of iterations for the QR method
+     * @param eigvals eigenvalues are stored here
+     * @param eigvecs eigenvectors are stored here
+     */
+    @Nonnull
+    public static void tridiagonalEigen(@Nonnull final RealMatrix T, @Nonnull final int nIter,
+            @Nonnull final double[] eigvals, @Nonnull final RealMatrix eigvecs) {
+        Preconditions.checkArgument(Arrays.deepEquals(T.getData(), T.transpose().getData()),
+                "Target matrix T must be a symmetric (tridiagonal) matrix");
+        Preconditions.checkArgument(eigvecs.getRowDimension() == eigvecs.getColumnDimension(),
+                "eigvecs must be a square matrix");
+        Preconditions.checkArgument(T.getRowDimension() == eigvecs.getRowDimension(),
+                "T and eigvecs must be the same shape");
+        Preconditions.checkArgument(eigvals.length == eigvecs.getRowDimension(),
+                "Number of eigenvalues and eigenvectors must be same");
+
+        int nEig = eigvals.length;
+
+        // initialize eigvecs as an identity matrix
+        eigvecs.setSubMatrix(new double[nEig][nEig], 0, 0);
+        for (int i = 0; i < nEig; i++) {
+            eigvecs.setEntry(i, i, 1);
+        }
+
+        RealMatrix T_ = T.copy();
+
+        for (int i = 0; i < nIter; i++) {
+            // QR decomposition for the tridiagonal matrix T
+            RealMatrix R = new Array2DRowRealMatrix(new double[nEig][nEig]);
```

Comment:
better to avoid copy in
https://github.com/apache/commons-math/blob/master/src/main/java/org/apache/commons/math4/linear/Array2DRowRealMatrix.java#L77

use `new Array2DRowRealMatrix(nEig, nEig)`